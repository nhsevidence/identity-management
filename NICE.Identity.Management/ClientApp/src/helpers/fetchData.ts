export const fetchData = async (url: string, options?: {}) => {
	let response, data;
	try {
		response = await fetch(url, options);
		data = await response.json();
	} catch (err) {
		let error: Error = err;
		return error;
	}

	if (response.status === 200) {
		return data;
	} else {
		let error = new Error(data.message);
		return error;
	}
};
