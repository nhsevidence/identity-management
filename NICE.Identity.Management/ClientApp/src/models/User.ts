
export type UserStatus = 'pending' | 'active' | 'incomplete'

export class User{
	id: string = ''
	firstName: string = ''
	lastName: string = ''
	email: string = ''
	status?: UserStatus = 'incomplete'
	locked?: boolean = false
}

export interface UserResponse {
	data: Array<User>
	count: number
}

export interface DeleteUserObj extends Partial<User>{
	id: string
}