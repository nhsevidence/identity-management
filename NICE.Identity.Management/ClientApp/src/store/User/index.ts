export * from './User'
export * from './UserProfile'