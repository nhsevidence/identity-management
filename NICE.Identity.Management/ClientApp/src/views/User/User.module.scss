.summaryList {
  display: flex;
  align-items: center;

  &Label {
    font-weight: bold;
    flex: 1;
  }
  &Detail {
    flex: 3;
    display: flex;
    align-items: center;
    justify-content: space-between;

    > * {
      margin: 0;
    }
  }
}
