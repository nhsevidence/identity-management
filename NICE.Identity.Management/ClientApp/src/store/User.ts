import {Store} from './'
import {User} from '../models/User'

export class UserStore extends Store<User>{
    constructor(){
        super()
    }

    // specific methods for the User store
}