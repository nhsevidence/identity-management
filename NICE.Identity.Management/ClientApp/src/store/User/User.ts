import { Store } from "../";
import { User } from "../../models/User";
import { url } from "../../constants/urls";

export class UserStore extends Store<User> {
  constructor() {
    super();
    this.url = url.users
  }
  // specific methods for the User store
}
