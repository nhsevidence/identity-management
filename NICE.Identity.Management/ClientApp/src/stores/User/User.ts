import { Store } from "../CommonStore"
import { User } from "../../models/User"
import { url } from "../../constants/urls"

export class UserStore extends Store<User> {
  constructor() {
    super(url.users)
  }
  // specific methods for the User store
}
