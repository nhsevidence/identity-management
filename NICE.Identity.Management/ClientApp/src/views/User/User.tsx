import React, { Component } from "react";

import { RouteComponentProps, Link, Redirect } from "react-router-dom";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";

import { UserType } from "../../models/types";
import { Endpoints } from "../../data/endpoints";
import { UnlockUser } from "../../components/UnlockUser/UnlockUser";
import { UserStatus } from "../../components/UserStatus/UserStatus";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";

import styles from "./User.module.scss";

type TParams = { id: string };

type UserProps = {} & RouteComponentProps<TParams>;

type UserState = {
	data: UserType;
	error?: Error;
	redirect: boolean;
	isLoading: boolean;
};

export class User extends Component<UserProps, UserState> {
	constructor(props: UserProps) {
		super(props);

		this.state = {
			data: {} as UserType,
			redirect: false,
			isLoading: true,
		};
	}

	handleError = (error: Error) => {
		this.setState({ error });
	};

	updateData = (updatedData: UserType) => {
		if (!Object.keys(updatedData).length) {
			this.setState({ redirect: true });
		} else {
			this.setState({ data: updatedData });
		}
	};

	fetchData = async (url: string) => {
		this.setState({ isLoading: true });

		let response, data;
		try {
			response = await fetch(url);
			data = await response.json();
		} catch (err) {
			let error: Error = err;

			this.setState({ error, isLoading: false });
			return;
		}

		this.setState({ isLoading: false });

		if (response.status === 200) {
			this.setState({ data });
		} else {
			this.setState({
				error: new Error(data.message),
			});
		}
	};

	componentDidMount() {
		this.fetchData(Endpoints.user(this.props.match.params.id));
	}

	render() {
		const { data, error, redirect, isLoading } = this.state;

		if (redirect) {
			return <Redirect to="/users" />;
		}

		let lastBreadcrumb = `${data.firstName} ${data.lastName}`;

		if (isLoading) {
			lastBreadcrumb = "Loading user details";
		}
		if (error) {
			lastBreadcrumb = "Error";
		}

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb to="/users" elementType={Link}>
						Users
					</Breadcrumb>
					<Breadcrumb>{lastBreadcrumb}</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader
							preheading="User profile for"
							heading={
								isLoading
									? "User details"
									: `${data.firstName} ${data.lastName}`
							}
						/>
						<Grid>
							<GridItem cols={12} md={9} aria-busy={isLoading}>
								{isLoading ? (
									<p>Loading...</p>
								) : (
									<>
										<div className={`${styles.summaryList} pv--c`}>
											<span className={styles.summaryListLabel}>
												Account information
											</span>
											<div className={styles.summaryListDetail}>
												<UserStatus user={data} />

												<UnlockUser
													id={data.userId}
													isLocked={data.isLockedOut}
													onToggleLock={this.updateData}
													onError={this.handleError}
												/>
											</div>
										</div>

										<div className={`${styles.summaryList} pv--c mb--d`}>
											<span className={styles.summaryListLabel}>
												Email address
											</span>
											<span className={styles.summaryListDetail}>
												{data.emailAddress}
											</span>
										</div>

										<hr className="mv--b" />

										<h2 className="h3">Permanently delete this account</h2>
										<p>
											The account will no longer be available, and all data in
											the account will be permanently deleted.
										</p>
										<Link
											data-qa-sel="delete-user-link"
											to={`/users/${data.userId}/delete`}
										>
											Delete user
										</Link>
									</>
								)}
							</GridItem>
						</Grid>
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
