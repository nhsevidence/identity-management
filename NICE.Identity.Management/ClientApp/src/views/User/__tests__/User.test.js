import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import fetchMock from "fetch-mock";
import toJson from "enzyme-to-json";

import { User } from "../User";
import singleUser from "./singleUser.json";
import { nextTick } from "../../../utils/nextTick";

import * as fetchData from "../../../helpers/fetchData";

describe("User", () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	afterEach(fetchMock.reset);

	it("should show loading message before data has been loaded", () => {
		fetchMock.get("*", {});
		const wrapper = shallow(<User match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetchData during componentDidMount", () => {
		fetchMock.get("*", {});
		const wrapper = shallow(<User match={match} />);
		const instance = wrapper.instance();
		jest.spyOn(fetchData, "fetchData");
		instance.componentDidMount();
		expect(fetchData.fetchData).toHaveBeenCalledTimes(1);
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetchMock.get("*", singleUser);
		const wrapper = shallow(<User match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetch returns 401 error", async () => {
		fetchMock.get("*", 401);
		const wrapper = mount(
			<MemoryRouter>
				<User match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetch returns 500 error", async () => {
		fetchMock.get("*", 500);
		const wrapper = mount(
			<MemoryRouter>
				<User match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});
});
