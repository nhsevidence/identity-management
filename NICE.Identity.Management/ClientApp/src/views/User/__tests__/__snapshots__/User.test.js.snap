// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot after data has been loaded 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={[Function]}
      to="/"
    >
      Home
    </Breadcrumb>
    <Breadcrumb
      elementType={[Function]}
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Vernita Green
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="Vernita Green"
  />
  <Grid
    elementType="div"
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      aria-busy={false}
      cols={12}
      elementType="div"
      md={9}
    >
      <Panel
        variant="supporting"
      >
        <UserStatus
          user={
            Object {
              "blocked": true,
              "email_address": "vernita.green@gmail.com",
              "email_verified": true,
              "first_name": "Vernita",
              "id": 2,
              "last_name": "Green",
              "user_id": "usr_5457edea1b8f33391a000002",
            }
          }
        />
        <p>
          User: 
          Vernita
           
          Green
        </p>
        <UnlockUser
          id={2}
          isLocked={true}
          onError={[Function]}
          onToggleLock={[Function]}
        />
        <Link
          to="/users/2/delete"
        >
          Delete user
        </Link>
      </Panel>
    </GridItem>
  </Grid>
</Fragment>
`;

exports[`should show error message when fetch returns 401 error 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={[Function]}
      to="/"
    >
      Home
    </Breadcrumb>
    <Breadcrumb
      elementType={[Function]}
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Error
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="Error"
  />
  <ErrorMessage
    error={[FetchError: invalid json response body at / reason: Unexpected end of JSON input]}
  />
</Fragment>
`;

exports[`should show error message when fetch returns 500 error 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={[Function]}
      to="/"
    >
      Home
    </Breadcrumb>
    <Breadcrumb
      elementType={[Function]}
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Error
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="Error"
  />
  <ErrorMessage
    error={[FetchError: invalid json response body at / reason: Unexpected end of JSON input]}
  />
</Fragment>
`;
